empress/support_files/js/layouts-util.js

```diff
@@ -0,0 +1,499 @@
+     *                   -xCoords
+     *                   -yCoords
+     *                   -highestChildYr
+     *                   -lowestChildYr
+     *                   -yScalingFactor
+     *                  Each of these properties (except for yScalingFactor)
+     *                  maps to an Array where data for each node is stored in
+     *                  postorder. yScalingFactor maps to a Number.
+     */
+    function rectangularLayout(tree, width, height, normalize = true) {
+        // NOTE: This doesn't draw a horizontal line leading to the root "node"
+        // of the graph. See https://github.com/biocore/empress/issues/141 for
+        // context.
+
+        var maxWidth = 0;
+        var maxHeight = 0;
+        var prevY = 0;
+        var xCoord = new Array(tree.size + 1).fill(0);
+        var yCoord = new Array(tree.size + 1).fill(0);
+        var highestChildYr = new Array(tree.size + 1).fill(0);
+        var lowestChildYr = new Array(tree.size + 1).fill(
+            Number.POSITIVE_INFINITY
+        );
+
+        // postorder
+        for (var i = 1; i <= tree.size; i++) {
+            if (tree.isleaf(tree.postorderselect(i))) {
+                yCoord[i] = prevY;
+                prevY += 1;
+                if (yCoord[i] > maxHeight) {
+                    maxHeight = yCoord[i];
+                }
+            } else {
+                // Center internal nodes above their children
+                // We could also center them above their tips, but (IMO) this
+                // looks better ;)
+                var sum = 0;
+                var numChild = 0;
+                var child = tree.fchild(tree.postorderselect(i));
+                while (child !== 0) {
+                    child = tree.postorder(child);
+                    sum += yCoord[child];
+                    child = tree.nsibling(tree.postorderselect(child));
+                    numChild++;
+                }
+                yCoord[i] = sum / numChild;
+            }
+        }
+
+        // iterates in preorder
+        var parent;
+        for (i = 2; i <= tree.size; i++) {
+            var node = tree.postorder(tree.preorderselect(i));
+            parent = tree.postorder(tree.parent(tree.preorderselect(i)));
+
+            xCoord[node] = xCoord[parent] + tree.length(tree.preorderselect(i));
+            if (maxWidth < xCoord[node]) {
+                maxWidth = xCoord[node];
+            }
+        }
+
+        // We don't check if max_width == 0 here, because we check when
+        // constructing an Empress tree that it has at least one positive
+        // branch length and no negative branch lengths. (And if this is the
+        // case, then max_width must be > 0.)
+        var xScalingFactor = width / maxWidth;
+
+        // Since this will be multiplied by 0 for every node, we can set
+        // this to any real number and get the intended "effect" of keeping
+        // every node's y-coordinate at 0.
```

Comment:
```suggestion
        // If the tree is a straight line (i.e. all internal nodes have exactly
        // one child), every node in the tree will have a y-coordinate of 0. In
        // this case, this scaling factor won't matter, since it'll get
        // multiplied by 0 for every node.
```

eesh, I had no idea what this comment meant at first. this is probably a problem since IIRC i wrote it :P this should make it clearer, hopefully...